wallet: Never schedule MaybeCompactWalletDB when -flushwallet is off

maflcko opened the pull request:
User-facing, this is a refactor. Internally, the scheduler does not have to call a mostly empty function every half a second.

MarcoFalke commented:
Can be tested by applying the following diff and observing a *failing* test:

```diff
diff --git a/test/functional/wallet_dump.py b/test/functional/wallet_dump.py
index 6bfb468823..f30589761a 100755
--- a/test/functional/wallet_dump.py
+++ b/test/functional/wallet_dump.py
@@ -190,7 +190,7 @@ class WalletDumpTest(BitcoinTestFramework):
         assert_raises_rpc_error(-8, "already exists", lambda: self.nodes[0].dumpwallet(wallet_enc_dump))
 
         # Restart node with new wallet, and test importwallet
-        self.restart_node(0, ['-wallet=w2'])
+        self.restart_node(0, ['-wallet=w2', '-flushwallet=0'])
 
         # Make sure the address is not IsMine before import
         result = self.nodes[0].getaddressinfo(multisig_addr)

Commit:
gui tests: Limit life-time of dummy testing setup

Its only purpose is to create a directory. So instead of keeping it
alive, use it only to get the path of the directory and then create it
explicitly.

Commit:
wallet: Pass unused args to StartWallets

This refactor does not change behavior

Commit:
wallet: Never schedule MaybeCompactWalletDB when -flushwallet is off

Commit:
test: Add smoke test to check that wallets are flushed by default

MarcoFalke commented:
Rebased

ryanofsky commented on src/interfaces/wallet.cpp:
```diff
@@ -491,7 +491,7 @@ class WalletClientImpl : public ChainClient
     bool verify() override { return VerifyWallets(m_chain, m_wallet_filenames); }
     bool load() override { return LoadWallets(m_chain, m_wallet_filenames); }
-    void start(CScheduler& scheduler) override { return StartWallets(scheduler); }
+    void start(CScheduler& scheduler, const ArgsManager& argsman) override { return StartWallets(scheduler, argsman); }
```
In commit "wallet: Pass unused argsman to StartWallets()" (faccdaba583288273399481b1f7892a910080fda)

Not a big deal and I can fix it later, but this is a bit broken with #10102 because it is trying to pass a whole argsman object from the node to wallet process just to pass a single bool. (See "Interface method parameter and return types should either be serializable or be other interface classes" examples in https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#internal-interface-guidelines)

Recommended fix for this would be to pass `-flushwallet` bool setting to StartWallets the same way the `-wallet` list setting gets passed to LoadWallets above: first read in [`WalletInit::Construct`](https://github.com/bitcoin/bitcoin/blob/73407ff65cefa29ce522cb3cb43898497e8afa30/src/wallet/init.cpp#L120) and passed as an argument to `MakeWalletClient`, stored in the WalletClientImpl instance, and used as needed

MarcoFalke replied:
Thanks, tried to implement your suggestion in branch 9ddddd2d927b3a752e4698c4771fc68cdc2d9698

ryanofsky replied:
> Thanks, tried to implement your suggestion in branch [9ddddd2](https://github.com/bitcoin/bitcoin/commit/9ddddd2d927b3a752e4698c4771fc68cdc2d9698)

Branch looks good. Current version is fine too in case you prefer that or the suggestion doesn't work.

 I was thinking maybe the best another approach after #19096 would be to add an ArgsManager pointer to the WalletContext struct, and initializing it from a new `MakeWalletClient` `ArgsManager&` argument so the wallet filenames and periodic flush arguments and temporary storage could be dropped

MarcoFalke replied:
> add an ArgsManager pointer to the WalletContext struct

thx, done

ryanofsky reviewed (approved):
Code review ACK fa6195467a3533fcdcc001d559555a86717cd5d3

ryanofsky reviewed (approved):
Code review ACK 9ddddd2d927b3a752e4698c4771fc68cdc2d9698. SInce last review was changed to pass periodic flush option from WalletInit::Construct

ryanofsky commented on src/wallet/test/init_test_fixture.h:
```diff
@@ -18,7 +18,6 @@ struct InitWalletDirTestingSetup: public BasicTestingSetup {
     fs::path m_datadir;
     fs::path m_cwd;
     std::map<std::string, fs::path> m_walletdir_path_cases;
-    NodeContext m_node;
```
In commit "test: Avoid overwriting the NodeContext member of the testing setup" (fa00c37cba98794c1c12c4b77e885d831d9bf183)

Wow, good catches. I wonder if there was a compiler warning we could enable that would have caught these.

ryanofsky commented on src/util/check.h:
```diff
@@ -42,4 +42,12 @@ class NonFatalCheckError : public std::runtime_error
 
+/** Assert that the passed pointer is not nullptr and return a reference to the pointed object */
+template <typename PtrT>
+typename std::remove_reference<decltype(*PtrT{})>::type& Ensure(const PtrT& ptr)
```
In commit "wallet: Pass unused args to WalletClientImpl" (fa38755d2fc444589ae506e4edd8597ab25e55e9)

Slight preference for calling this `AssertNotNull` instead of `Ensure` just because it could wind up being used widely, and it'd be nice for it to be obvious what it does.

Also for code review purposes, C++ programmers are already pretty trained to know that `*ptr` is dangerous and can kill the program, while `Ensure(ptr)` might look more innocuous if it's part of a larger expression. And since the name doesn't contain "Assert" or "Abort", it might actually be clearer if this returned a pointer so the caller is forced to write `*`. Could also go more generic and provide

```c++
template<typename T> T&& Assert(T&& value) { assert(value); return std::forward<T>(value); }
```

ryanofsky reviewed (approved):
Code review ACK 00000d48588689f7576a8f4cbd9eb8edee3b53b0. No important suggestions. First two commits are the same since last review, last commit is basically the same, middle commits switch back to passing ArgsManager instead of the bool, but going through WalletContext struct from #19096

fanquake commented:
Concept ACK

MarcoFalke commented:
commit id 00000d4 was lost in a rebase

ryanofsky reviewed (approved):
Code review ACK faf18d6c86cbb3e90070a971ee341cc76fb1a8f5. Changes since previous review were replacing Ensure with Assert, and adding new checks to ensure the NDEBUG is disabled and that assert isn't called from RPC code. First two commits are new, previous shadow fix commit became #19188, and the other commits are basically unchanged except being divided differently and replacing assert with ensure. Long list. Nice cleanups, though

ryanofsky commented on src/util/check.h:
```diff
@@ -46,4 +46,12 @@ class NonFatalCheckError : public std::runtime_error
 
+/** Identity function. Abort if the value compares equal to zero */
+template <typename T>
+T&& Assert(T&& value)
```
In commit "util: Add Assert identity function" (fa54cf107319609ff44de960b987bd09b46945d0)

Thinking about this, I think it would be better to change signature to `T Assert(T&& value)` to avoid a possibility of returning a reference to a temporary. Contrived example but I think `p` would be an invalid reference here returning `T&&` instead of `T`:

```c++
std::unique_ptr<int>&& p = Assert(MakeUnique<int>(5));
```

MarcoFalke commented:
I've split out the first commits into #19277

ryanofsky reviewed (approved):
Would be good to merge #19277 first, but code review ACK fac173681bfcd0debd7f203aeb4eb6910407cce8. Only change since last review is suggested change to assert return type

MarcoFalke commented:
Rebased

promag commented on src/wallet/bdb.cpp:
```diff
@@ -623,8 +623,8 @@ bool BerkeleyDatabase::PeriodicFlush()
-    for (auto it = env->mapFileUseCount.begin() ; it != env->mapFileUseCount.end(); it++) {
-        if ((*it).second > 0) return false;
+    for (const auto& use_count : env->mapFileUseCount) {
+        if (use_count.second > 0) return false;
```
fa73493930e35850e877725167dc9d42e47015c8

nit, if you feel like cleaning it a bit more maybe do this below:
```cpp
if (env->mapFileUseCount.erase(strFile) == 0) return false
```
And drop `.erase(it)`.

MarcoFalke replied:
Thanks, but I'll leave that for a follow-up

Commit:
refactor: Use C++11 range-based for loop

jnewbery reviewed (commented):
utACK fa73493930e35850e877725167dc9d42e47015c8

One comment on the new test. Feel free to ignore.

MarcoFalke commented on test/functional/wallet_dump.py:
```diff
@@ -202,5 +202,10 @@ def run_test(self):
         assert result['ismine']
 
+        self.log.info('Check that wallet is flushed')
+        with self.nodes[0].assert_debug_log(['Flushing wallet.dat'], timeout=20):
```
I think both can be done at the same time, but asserting the modification time increases requires a `time.sleep(1)`, since `os.path.getmtime` returns seconds. I don't like sleeps longer than 200 ms in tests, so I'll leave that for a follow-up ;)

jnewbery replied:
No need for a follow-up. Fast tests are nice too!

meshcollider reviewed (commented):
utACK fa73493930e35850e877725167dc9d42e47015c8

ryanofsky reviewed (approved):
Code review ACK fa73493930e35850e877725167dc9d42e47015c8. Just rebased since last review
